src/main/java/io/cryostat/rules/MatchExpressionValidator.java

```diff
@@ -0,0 +1,254 @@
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ * The above copyright notice and either this complete permission notice or at
+ * a minimum a reference to the UPL must be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.rules;
+
+import jdk.nashorn.api.tree.BreakTree;
+import jdk.nashorn.api.tree.CaseTree;
+import jdk.nashorn.api.tree.CatchTree;
+import jdk.nashorn.api.tree.ClassDeclarationTree;
+import jdk.nashorn.api.tree.ClassExpressionTree;
+import jdk.nashorn.api.tree.CompilationUnitTree;
+import jdk.nashorn.api.tree.ContinueTree;
+import jdk.nashorn.api.tree.DebuggerTree;
+import jdk.nashorn.api.tree.DoWhileLoopTree;
+import jdk.nashorn.api.tree.ErroneousTree;
+import jdk.nashorn.api.tree.ExportEntryTree;
+import jdk.nashorn.api.tree.ForInLoopTree;
+import jdk.nashorn.api.tree.ForLoopTree;
+import jdk.nashorn.api.tree.ForOfLoopTree;
+import jdk.nashorn.api.tree.FunctionCallTree;
+import jdk.nashorn.api.tree.FunctionDeclarationTree;
+import jdk.nashorn.api.tree.FunctionExpressionTree;
+import jdk.nashorn.api.tree.ImportEntryTree;
+import jdk.nashorn.api.tree.InstanceOfTree;
+import jdk.nashorn.api.tree.LabeledStatementTree;
+import jdk.nashorn.api.tree.ModuleTree;
+import jdk.nashorn.api.tree.NewTree;
+import jdk.nashorn.api.tree.Parser;
+import jdk.nashorn.api.tree.RegExpLiteralTree;
+import jdk.nashorn.api.tree.ReturnTree;
+import jdk.nashorn.api.tree.SimpleTreeVisitorES5_1;
+import jdk.nashorn.api.tree.SpreadTree;
+import jdk.nashorn.api.tree.ThrowTree;
+import jdk.nashorn.api.tree.Tree;
+import jdk.nashorn.api.tree.TreeVisitor;
+import jdk.nashorn.api.tree.TryTree;
+import jdk.nashorn.api.tree.UnaryTree;
+import jdk.nashorn.api.tree.WhileLoopTree;
+import jdk.nashorn.api.tree.WithTree;
+import jdk.nashorn.api.tree.YieldTree;
+
+public class MatchExpressionValidator {
+
+    private final Parser parser = Parser.create();
+
+    private final TreeVisitor<Void, String> treeVisitor =
+            new SimpleTreeVisitorES5_1<Void, String>() {
```

Comment:
I'm anonymously extending `SimpleTreeVisitorES5_1`, which is a concrete class implementing the `TreeVisitor` interface, and only overriding the visitor methods that we want to disallow. All other visitor methods are implemented by the `SimpleTreeVisitorES5_1` base class and inherited by this anonymous extension class, which implicitly pass those AST nodes by default. So, any visitor methods for node types not overridden by this class are implicitly passed, and all the ones I have explicitly named by overriding the relevant visitors are failed. Do you mean to replace this with a class where all of the visitor methods are directly implemented by the class and there is a composition delegate `SimpleTreeVisitor` field for passing, so that all the AST node types/visitor methods are enumerated in the implementation, rather than having the implicit passes that I have here?

Reply:
Right, so that nothing is implicitly passed. I'm thinking it might be slightly less error prone to maintain, but I'll leave it up to you.

Reply:
Okay, that's fair enough. I'll refactor it to do as you suggest. It's also probably worth pulling it into its own class, too.

Reply:
Hmm - this doesn't seem to work as expected, actually. The `TreeVisitor` only seems to have one of its methods called, and then execution resumes at the line after `cut.accept(treeVisitor, rule.getMatchExpression())`. I expected its methods to be called repeatedly as the parser proceeded, but that doesn't seem to be how it works. The first node type passed is the CompilationUnit, so it will pass everything after that point. I also tried iterating over `cut.getSourceElements()`, but this just gives me a list of AST children of the CompilationUnit, which can also be other generally acceptable abstract node types. So to get the "explicit passing to a delegate" piece working, it seems like I would need to recursively descend the whole AST myself - doesn't seem worth it.

I'll extract the TreeVisitor to a top-level class of its own at least, but I'll leave it with the original implementation of only observing fail-worthy AST nodes.